controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -0,0 +1,497 @@
+func NewEndpointsWatcher(k8sAPI *k8s.API, log *logging.Entry) *EndpointsWatcher {
+	ew := &EndpointsWatcher{
+		publishers: make(map[ServiceID]*servicePublisher),
+		k8sAPI:     k8sAPI,
+		log: log.WithFields(logging.Fields{
+			"component": "endpoints-watcher",
+		}),
+	}
+
+	k8sAPI.Pod().Informer().AddIndexers(cache.Indexers{podIPIndex: func(obj interface{}) ([]string, error) {
+		if pod, ok := obj.(*corev1.Pod); ok {
+			return []string{pod.Status.PodIP}, nil
+		}
+		return []string{""}, fmt.Errorf("object is not a pod")
+	}})
+
+	k8sAPI.Svc().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ew.addService,
+		DeleteFunc: ew.deleteService,
+		UpdateFunc: func(_, obj interface{}) { ew.addService(obj) },
+	})
+
+	k8sAPI.Endpoint().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ew.addEndpoints,
+		DeleteFunc: ew.deleteEndpoints,
+		UpdateFunc: func(_, obj interface{}) { ew.addEndpoints(obj) },
+	})
+
+	return ew
+}
+
+////////////////////////
+/// EndpointsWatcher ///
+////////////////////////
+
+// Subscribe to an authority.
+// The provided listener will be updated each time the address set for the
+// given authority is changed.
+func (ew *EndpointsWatcher) Subscribe(authority string, listener EndpointUpdateListener) error {
+	id, port, err := GetServiceAndPort(authority)
+	if err != nil {
+		return err
+	}
+	ew.log.Infof("Establishing watch on endpoint [%s:%d]", id, port)
+
+	ew.publishersMu.Lock()
+	// If the service doesn't yet exist, create a stub for it so the listener can
+	// be registered.
+	sp, ok := ew.publishers[id]
+	if !ok {
+		sp = ew.newServicePublisher(id)
+		ew.publishers[id] = sp
+	}
+	ew.publishersMu.Unlock()
+
+	sp.subscribe(port, listener)
+	return nil
+}
+
+// Unsubscribe removes a listener from the subscribers list for this authority.
+func (ew *EndpointsWatcher) Unsubscribe(authority string, listener EndpointUpdateListener) {
+	id, port, err := GetServiceAndPort(authority)
+	if err != nil {
+		ew.log.Errorf("Invalid service name [%s]", authority)
+		return
+	}
+	ew.log.Infof("Stopping watch on endpoint [%s:%d]", id, port)
+
+	ew.publishersMu.RLock()
+	sp, ok := ew.publishers[id]
+	ew.publishersMu.RUnlock()
+	if !ok {
+		ew.log.Errorf("Cannot unsubscribe from unknown service [%s:%d]", id, port)
+	}
+	sp.unsubscribe(port, listener)
+}
+
+func (ew *EndpointsWatcher) addService(obj interface{}) {
+	service := obj.(*corev1.Service)
+	id := ServiceID{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+	}
+
+	ew.publishersMu.Lock()
+	sp, ok := ew.publishers[id]
+	if !ok {
+		sp = ew.newServicePublisher(id)
+		ew.publishers[id] = sp
+	}
+	ew.publishersMu.Unlock()
+
+	sp.updateService(service)
+}
+
+func (ew *EndpointsWatcher) deleteService(obj interface{}) {
+	service := obj.(*corev1.Service)
+	id := ServiceID{
+		Namespace: service.Namespace,
+		Name:      service.Name,
+	}
+
+	ew.publishersMu.RLock()
+	sp, ok := ew.publishers[id]
```

Comment:
I notice that we never delete `ew.publishers[id]`. What is the rationale behind this design?

Reply:
We could delete the publisher once the number of subscribers drops to 0, but note that it would be re-created as soon as there is an update to that endpoints.  We also want to allow subscriptions to services that don't exist, so that clients can subscribe and wait for a service to be created.

Therefore we only really want to delete a publisher if it has no subscribers AND is for a service that does not currently exist.  Getting this logic exactly right may be tricky so I'd like to punt on this for now.  This is technically leaky, but should only matter when there is a high rate of service churn with new unique names.